src/test/shared/vscode/runCommand.test.ts

```diff
@@ -91,5 +91,113 @@ describe('runCommand', function () {
+                const command = Commands.register(
+                    {
+                        id: '_test.telemetryName.commandSourceArgDebounceCount',
+                        uniqueArgIndexes: [new IndexedAttribute(0, 'source')],
+                    },
+                    (source: string) => {}
+                )
+
+                const clock = installFakeClock()
+
+                await command.execute('a')
+                // following 3 events will be throttled
+                await command.execute('a')
+                await command.execute('a')
+                await command.execute('a')
+                // advance clock past throttle time
+                clock.tick(defaultTelemetryThrottleMs)
+                // will contain a 'debounceCount' of 3
+                await command.execute('a')
+
+                const events = globals.telemetry.logger.query({ metricName: 'vscode_executeCommand' })
+                assert.strictEqual(events.length, 2)
+
+                const firstEvent = events[0]
+                assert.strictEqual(firstEvent['source'], 'a')
+                assert.strictEqual(firstEvent['debounceCount'], undefined)
+
+                const secondEvent = events[1]
+                assert.strictEqual(secondEvent['source'], 'a')
+                assert.strictEqual(secondEvent['debounceCount'], '3')
+            })
+
+            it('throws if out of index', async function () {
+                const command = Commands.register(
+                    {
+                        id: '_test.telemetryName.throwsOnOutOfIndex',
+                        uniqueArgIndexes: [new IndexedAttribute(1, 'source')], // out of index
+                    },
+                    (source: any) => {}
+                )
+                await assert.rejects(() => command.execute(1), new ToolkitError(`Unique arg indexes exceed the # of args for: "_test.telemetryName.throwsOnOutOfIndex"`))
+            })
+
+            it('throws if arg type is not supported', async function () {
+                const command = Commands.register(
+                    {
+                        id: '_test.telemetryName.throwsOnUnsupportedType',
+                        uniqueArgIndexes: [0],
+                    },
+                    (source: any) => {}
+                )
+
+                // supported
+                await command.execute('stringsAreSupported')
+                await command.execute(true)
+
+                // unsupported
+                await assert.rejects(() => command.execute(1), new ToolkitError(`Cannot create unique key for type: "number"`))
+            })
+
+            it('throws if "source" is not a string', async function () {
+                const command = Commands.register(
+                    {
+                        id: '_test.telemetryName.throwsOnBadSourceType',
+                        uniqueArgIndexes: [new IndexedAttribute(0, 'source')],
```

Comment:
- if `IndexedAttribute` were an interface the ergnomics may improve. Classes lose some benefits of typescript's structural typing.
    - maybe we could flip it so that the names are specified as object keys, and the values are the indexes. 
        - that naturally also avoids duplicate indexes.
- `uniqueArgIndexes` is an example of a name that is trying to explain a lot but ends up being opaque anyway. Simple is usually better. Docstrings can reveal details if necessary, in an opt-in fashion.

```suggestion
                        key: {'source': 0, 'foo': 3},
```

Reply:
Good idea. Created interface `ArgIndexes` which does similar to what you mentioned.

The will instead provide a single object similar to `{0: 'source'}` which ensures unique indexes since the key can only be used once